File: WorkflowPatternFinder/Gensim/ProcessTreeLoader.py
import enum
from enum import Enum # learn more: https://python.org/pypi/enum
import sys
from ProcessTree import *
from Node import *
import xml.etree.ElementTree as ET

class NodeType(Enum):
    Xor = 1
    XorLoop = 2
    And = 3
    AndLoop = 4
    Sequence = 5
    SequenceLoop = 6
    ManualTask = 7
    Tau = 8

class ProcessTreeLoader(object):
    """This class loads in a process tree from a .ptml file."""
    @classmethod
    def LoadTree(self, filePath):
        ProcessTreeLoader.sourceFile = filePath
        tree = None
        doc = ET.parse(filePath)
        for child in doc.iter():
          if child.tag == 'processTree':
            tree = ProcessTree(filePath, child.attrib['id'], child.attrib['root'])
          elif child.tag == 'xor':
            tree.AddNode(Node(NodeType.Xor, child.attrib['id']))
          elif child.tag == 'xorLoop':
            tree.AddNode(Node(NodeType.XorLoop, child.attrib['id']))
          elif child.tag == 'and':
            tree.AddNode(Node(NodeType.And, child.attrib['id']))
          elif child.tag == 'andLoop':
            tree.AddNode(Node(NodeType.AndLoop, child.attrib['id']))
          elif child.tag == 'sequence':
            tree.AddNode(Node(NodeType.Sequence, child.attrib['id']))
          elif child.tag == 'sequenceLoop':
            tree.AddNode(Node(NodeType.SequenceLoop, child.attrib['id']))
          elif child.tag == 'manualTask':
            tree.AddNode(Node(NodeType.ManualTask, child.attrib['id'], child.attrib['name']))
          elif child.tag == 'automaticTask':
            tree.AddNode(Node(NodeType.Tau, child.attrib['id']))
          elif child.tag == 'parentsNode':
            parentId = child.attrib['sourceId']
            childId = child.attrib['targetId']
            tree.SetParentalRelation(parentId, childId)
        print('tree size:', tree.GetTreeSize())
        return tree  
            

File: WorkflowPatternFinder/WorkflowEventLogProcessor/WorkflowEventLogProcessor/WorkflowEventLogProcessor/ProcessTreeLoader.cs
using System;
using System.Xml;

namespace WorkflowEventLogProcessor
{
  public static class ProcessTreeLoader
  {
    public enum NodeType
    {
      Xor,
      XorLoop,
      And,
      AndLoop,
      Sequence,
      SequenceLoop,
      ManualTask,
      Tau
    };

    public static ProcessTree LoadTree(string filePath)
    {
      var sourceFile = filePath;
      ProcessTree tree = null;
      var settings = new XmlReaderSettings
      {
        DtdProcessing = DtdProcessing.Parse
      };
      var reader = XmlReader.Create(sourceFile, settings);
      reader.MoveToContent();
      while(reader.Read())
      {
        if(reader.IsStartElement())
        {
          switch(reader.Name)
          {
            case "processTree":
              {
                tree = new ProcessTree(filePath, reader["id"], reader["root"]);
                break;
              }
            case "xor":
              {
                tree?.AddNode(new Node(NodeType.Xor, reader["id"]));
                break;
              }
            case "xorLoop":
              {
                tree?.AddNode(new Node(NodeType.XorLoop, reader["id"]));
                break;
              }
            case "and":
              {
                tree?.AddNode(new Node(NodeType.And, reader["id"]));
                break;
              }
            case "andLoop":
              {
                tree?.AddNode(new Node(NodeType.AndLoop, reader["id"]));
                break;
              }
            case "sequence":
              {
                tree?.AddNode(new Node(NodeType.Sequence, reader["id"]));
                break;
              }
            case "sequenceLoop":
              {
                tree?.AddNode(new Node(NodeType.SequenceLoop, reader["id"]));
                break;
              }
            case "manualTask":
              {
                tree?.AddNode(new Node(NodeType.ManualTask, reader["id"], reader["name"]));
                break;
              }
            case "automaticTask":
              {
                tree?.AddNode(new Node(NodeType.Tau, reader["id"]));
                break;
              }
            case "parentsNode":
              {
                var parentId = new Guid(reader["sourceId"] ?? throw new InvalidOperationException());
                var childId = new Guid(reader["targetId"] ?? throw new InvalidOperationException());
                tree?.SetParentalRelation(parentId, childId);
                break;
              }
          }
        }
      }
      reader.Close();
      return tree;
    }
  }
}